src/za/ac/sun/cs/green/service/simplify/ConstantPropogation.java

```diff
@@ -0,0 +1,133 @@
+import za.ac.sun.cs.green.expr.IntConstant;
+import za.ac.sun.cs.green.expr.IntVariable;
+import za.ac.sun.cs.green.expr.Operation;
+import za.ac.sun.cs.green.expr.Variable;
+import za.ac.sun.cs.green.expr.Visitor;
+import za.ac.sun.cs.green.expr.VisitorException;
+
+public class ConstantPropogation extends BasicService {
+	
+	private int invocations = 0;
+
+	public ConstantPropogation(Green solver) {
+		super(solver);
+	}
+
+	@Override
+    public Set<Instance> processRequest(Instance instance) {
+        @SuppressWarnings("unchecked")
+        Set<Instance> result = (Set<Instance>) instance.getData(getClass());
+        if (result == null) {
+            final Map<Variable, Variable> map = new HashMap<Variable, Variable>();
+            final Expression e = simplify(instance.getFullExpression(), map);
+            final Instance i = new Instance(getSolver(), instance.getSource(), null, e);
+            result = Collections.singleton(i);
+            instance.setData(getClass(), result);
+        }
+        return result;
+    }
+
+    @Override
+    public void report(Reporter reporter) {
+        reporter.report(getClass().getSimpleName(), "invocations = " + invocations);
+    }
+
+	public Expression simplify(Expression expression, Map<Variable, Variable> map) {
+		try {
+			log.log(Level.FINEST, "Before Simplification: " + expression);
+			invocations++;
+			SimplifyVisitor simplifyVisitor = new SimplifyVisitor();
+			expression.accept(simplifyVisitor);
+			expression = simplifyVisitor.getExpression();
+			log.log(Level.FINEST, "After simplification: " + expression);
+		} catch (VisitorException x) {
+			log.log(Level.SEVERE, "encountered an exception -- this should not be happening!", x);
+		} finally {
+			return expression;
+		}		
+	}
+
+	public class SimplifyVisitor extends Visitor {
+		private Stack<Expression> stack;
+		private Map<IntVariable, IntConstant> map;
+
+		public SimplifyVisitor() {
+			stack = new Stack<Expression>();
+			map = new TreeMap<IntVariable, IntConstant>();
+		}
+
+		public Expression getExpression() {
+			return stack.pop();
+		}
+
+		@Override
+		public void postVisit(IntConstant constant) {
+			stack.push(constant);
+		}
+
+		@Override
+		public void preVisit(Operation operation) {
+			Operation.Operator op = operation.getOperator();
+			if (op.equals(Operation.Operator.EQ)) {
+				Expression opL = operation.getOperand(0);
+				Expression opR = operation.getOperand(1);
+				if ((opL instanceof IntConstant) && (opR instanceof IntVariable)) {
+					map.put((IntVariable) opR, (IntConstant) opL);
+				} else if ((opL instanceof IntVariable) && (opR instanceof IntConstant)) {
+					map.put((IntVariable) opL, (IntConstant) opR);
+				}
+			}
+		}
+
+		@Override
+		public void postVisit(IntVariable variable) {
+			stack.push(variable);
+		}
+
+		@Override
+		public void postVisit(Operation operation) {
+			Operation.Operator op = operation.getOperator();
+
+			if (stack.size() >= 2) {
```

Comment:
Same thing here, with regards to comments.

Reply:
Thanks Phillip, will work on that now